Readme.md

```diff
@@ -0,0 +1,1432 @@
+
+```scala
+case object EmptyList
+```
+
+Second, an ADT may combine several mutually exclusive variants of products:
+
+```scala
+sealed trait OptionalInt
+case class SomeInt(value: Int) extends OptionalInt
+case object NoneInt extends OptionalInt
+```
+
+Here `OptionalInt` is the ADT, and `SomeInt` and `NoneInt` are its variants. Each variant is a product ADT in itself. Such combinations of product types are called sum types, or sometimes disjoint unions. Another common example of such an ADT is a tree, which also demonstrates how an ADT can be recursive:
+
+```scala
+sealed trait Tree
+object Tree {
+  case class Node(left: Tree, right: Tree) extends Tree
+  case class Leaf(value: Int) extends Tree
+}
+```
+
+As can be seen above, in Scala sealed traits and case classes are used to create algebraic data types.
+
+ADTs are useful to model a wide range data structures, simple data type and state machines being the most prominent examples. Pattern matching together with the exhaustiveness analysis allow writing more correct code. Many algorithms on ADTs are naturally recursive, and pattern matching leads to very clear and correct code:
+
+```scala
+def findMax(tree: Tree): Int = tree match {
+  case Tree.Node(left, right) => Vector(findMax(left), findMax(right)).max
+  case Tree.Leaf(value) => value
+}
+```
+
+Because of the exhaustiveness analysis, such pattern matches will make sure that you didn't forget to check any variants of an ADT.
+
+#### Sum types declarations
+
+If you need to write a sum data type, declare all its variants inside the companion object of the root sealed trait:
+
+```scala
+// write this
+sealed trait IntList
+object IntList {
+  case class Cons(value: Int, rest: IntList) extends IntList
+  case object Empty extends IntList
+}
+
+// not this
+sealed trait IntList
+case class IntListCons(value: Int, rest: IntList) extends IntList
+case object IntListEmpty extends IntList
+```
+
+Putting variants of a sum type into the companion object reduces the namespace pollution, allows to use nicer names for variants and makes the code which uses these names clearer because it becomes visible to which type these variants belong.
+
+#### Enumerations
+
+One important case of sum ADTs are enumerations, i.e. a fixed small set of values (objects). Lots of things are modeled with enumerations, for example, state of some system:
+
+```scala
+sealed trait State
+object State {
+  case object Stopped extends State
+  case object Started extends State
+  case object Failed extends State
+}
+```
+
+So, when you need to declare something which would be an `enum` in Java, use the above pattern.
+
+Just as with any other ADT, declare all enumeration constants inside the companion object of the respective sealed trait. Do not use `scala.Enumeration` because it has very unnatural API.
+
+### Options and `null`
```

Comment:
I would suggest to elaborate some naming convention for fields of ``Option`` type.
I mean if there is some ``case class Foo(..., user: User, ...)`` it looks strange to have in a neighbour class something like ``case class Bar(..., user: Option[User], ...)``, because
```scala
val foo: Foo = ...
val bar: Bar = ...
foo.user.name // OK, but
bar.user.map(_.name).getOrElse("default") // Huh?
```
If there was some difference in names, it would look more natural.
Moreover, the typical naming convention for collections is to use plural nouns, e.g.:
``case class Dar(..., users: Seq[User], ...)``
So, why not to accept some naming convention for ``Option``s?
I saw two approaches: the ``Opt`` suffix and the ``_?`` suffix.
For me, the latter is a bit more self-descriptive because it more clearly separates
the name part from the option marker:
```scala
case class Foo(user: User)
case class Bar(user_? : Option[User])

foo.user.name
bar.user_?.map(_.name).getOrElse("default")
```
but I do not insist on it.